### PR TITLE
Check exit code for test subprocess

### PR DESCRIPTION
**Pull request by Hixie**
Refactor the test harness to handle failures better.

I still want to write a test that verifies that the harness catches errors when the engine crashes, so this is not complete, but I'm putting it out for review now.

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 1]
```diff
@@ -20,125 +19,184 @@ import '../dart/package_map.dart';
...
+
+    printTrace('ending test: $testPath');
+
+    CoverageCollector.instance.collectCoverage(
+      host: _kHost.address,
+      port: observatoryPort,
+      processToKill: process, // This kills the subprocess whether coverage is enabled or not.
+    );
+
+    temporaryDirectory.deleteSync(recursive: true);
```
Comment (abarth) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: I'd put this in a "finally" block to make it harder to leak temp directories.

**Comment by abarth** [comment]
LGTM

**Comment by eseidelGoogle** [comment]
Do we expect this to help with https://github.com/flutter/flutter/issues/5694 ?

**Comment by eseidelGoogle** [comment]
https://github.com/flutter/flutter/issues/2698 might be a better bug, there are a few relating to flutter test troubles on file.

**Comment by Hixie** [comment]
It'll definitely fix the first, not sure about the second but likely yes also.

**Comment by Hixie** [comment]
BTW the Travis red is because this is now catching the regression that started this whole thing.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: done

**Comment by jason-simmons** [comment]
LGTM

This is a lot easier to understand now

**Comment by Hixie** [comment]
Ok this has tests now and is ready for thorough review. Unless I hear otherwise, though, I'm gonna take Jason's LGTM and land it once Travis goes green. :-)

**Pull request merged**